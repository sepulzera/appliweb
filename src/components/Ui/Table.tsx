import * as React from 'react';

import MuiTable from '@material-ui/core/Table';
import { TableHead, TableRow, TableCell, TableBody } from '@material-ui/core';

/** Column definition for {@link Table}. */
export type column = {
  /** Technical name of the column. Should be a property of data. */
  name: string;
  /** Displayed title of the column. Should be localized. */
  title: string;
}

/** {@link Table} Props. */
export interface ITableProps {
  /** Caption for the table. */
  caption:    string;
  /** Column definition, see {@link column}. */
  columns:    Array<column>;
  /** Data to be displayed. Should include all properties defined by columns or it will throw. */
  data:       Array<any>;
}

/**
 * Renders an HTML table.
 *
 * @param props - {@link ITableProps}.
 */
const Table: React.FC<ITableProps> = (props: ITableProps) => {
  const cols: Array<string> = props.columns.map((col: column) => (
    col.name
  ));

  const getData = (dat: any, col: string) => {
    const d = dat[col];
    return d;
  };

  if (props.columns.length === 0) return null;

  return (
    <MuiTable size='small'>
      <caption>{props.caption}</caption>
      <TableHead>
        <TableRow>
          {props.columns.map((col: column) => (
            <TableCell>{col.title}</TableCell>
          ))}
        </TableRow>
      </TableHead>
      <TableBody>
        {props.data.map((dat: any) => (
          <TableRow key={dat.id}>
            {cols.map((col: string) => (
              <TableCell>{getData(dat, col)}</TableCell>
            ))}
          </TableRow>
        ))}
      </TableBody>
    </MuiTable>
  );
};

export default Table;
