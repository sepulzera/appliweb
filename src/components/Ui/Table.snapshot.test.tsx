/* eslint-disable jsx-a11y/accessible-emoji */
import React from 'react';
import renderer from 'react-test-renderer';

import Table from './Table';

const caption = 'This is a caption';

const columns = [
  {
    name: 'SomeProperty',
    title: 'Some Title',
  },
  {
    name: 'AnotherProperty',
    title: 'Another Title',
  },
];

const data = [
  {
    id: 'SomeProperty',
    SomeProperty:    'Some Simple value',
    AnotherProperty: <i>Some complex value</i>,
  },
  {
    id: 'SomeValue',
    SomeProperty:    <i>Another complex value</i>,
    AnotherProperty: 'Another simple value',
  },
];

test('render empty table with no columns', () => {
  const tree = renderer.create(<Table caption={caption} columns={[]} data={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render empty table', () => {
  const tree = renderer.create(<Table caption={caption} columns={columns} data={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render filled table', () => {
  const tree = renderer.create(<Table caption={caption} columns={columns} data={data} />).toJSON();
  expect(tree).toMatchSnapshot();
});
