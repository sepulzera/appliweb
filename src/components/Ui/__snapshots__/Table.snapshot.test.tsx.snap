// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render empty table 1`] = `
<table
  className="MuiTable-root"
  role={null}
>
  <caption>
    This is a caption
  </caption>
  <thead
    className="MuiTableHead-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root MuiTableRow-head"
      role={null}
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
        scope="col"
      >
        Some Title
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
        scope="col"
      >
        Another Title
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root"
    role={null}
  />
</table>
`;

exports[`render empty table with no columns 1`] = `null`;

exports[`render filled table 1`] = `
<table
  className="MuiTable-root"
  role={null}
>
  <caption>
    This is a caption
  </caption>
  <thead
    className="MuiTableHead-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root MuiTableRow-head"
      role={null}
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
        scope="col"
      >
        Some Title
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
        scope="col"
      >
        Another Title
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
      >
        Some Simple value
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
      >
        <i>
          Some complex value
        </i>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
      >
        <i>
          Another complex value
        </i>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
      >
        Another simple value
      </td>
    </tr>
  </tbody>
</table>
`;
